File: flipt-client-swift/Tests/FliptClientTests/FliptClientTests.swift
@testable import FliptClient
import XCTest

class FliptClientTests: XCTestCase {
    var evaluationClient: FliptClient?
    var fliptUrl: String = ""
    var authToken: String = ""

    override func setUp() {
        super.setUp()

        guard let fliptUrl = ProcessInfo.processInfo.environment["FLIPT_URL"],
              let authToken = ProcessInfo.processInfo.environment["FLIPT_AUTH_TOKEN"]
        else {
            XCTFail("FLIPT_URL and FLIPT_AUTH_TOKEN must be set")
            return
        }

        do {
            var tlsConfig: TlsConfig?

            let caCertPath = ProcessInfo.processInfo.environment["FLIPT_CA_CERT_PATH"]
            if let caCertPath, !caCertPath.isEmpty {
                tlsConfig = try TlsConfig.builder()
                    .caCertFile(caCertPath)
                    .insecureSkipHostnameVerify(true)
                    .build()
            } else {
                // Fallback to insecure for local testing
                tlsConfig = try TlsConfig.builder()
                    .insecureSkipVerify(true)
                    .build()
            }

            evaluationClient = try FliptClient(
                namespace: "default",
                url: fliptUrl,
                authentication: .clientToken(authToken),
                tlsConfig: tlsConfig)
        } catch {
            XCTFail("Failed to initialize EvaluationClient: \(error)")
        }
    }

    override func tearDown() {
        evaluationClient?.close()
        evaluationClient = nil
        super.tearDown()
    }

    func testEmptyFlagKey() {
        do {
            _ = try evaluationClient?.evaluateBoolean(
                flagKey: "",
                entityID: "someentity",
                evalContext: ["fizz": "buzz"])
            XCTFail("Expected an error, but got none")
        } catch let error as FliptClient.ClientError {
            XCTAssertFalse(error.localizedDescription.isEmpty)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testEmptyEntityId() {
        do {
            _ = try evaluationClient?.evaluateBoolean(flagKey: "flag1", entityID: "", evalContext: ["fizz": "buzz"])
            XCTFail("Expected an error, but got none")
        } catch let error as FliptClient.ClientError {
            XCTAssertFalse(error.localizedDescription.isEmpty)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testInvalidAuthentication() {
        do {
            let client = try FliptClient(
                namespace: "default",
                url: fliptUrl,
                authentication: .clientToken("invalid"))

            _ = try client.evaluateVariant(flagKey: "flag1", entityID: "someentity", evalContext: ["fizz": "buzz"])
            XCTFail("Expected an error, but got none")
        } catch let error as FliptClient.ClientError {
            XCTAssertFalse(error.localizedDescription.isEmpty)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testVariant() {
        do {
            let variantResult = try evaluationClient?.evaluateVariant(
                flagKey: "flag1",
                entityID: "someentity",
                evalContext: ["fizz": "buzz"])
            guard let variantResult else {
                XCTFail("variantResult is nil")
                return
            }
            XCTAssertTrue(variantResult.match)
            XCTAssertEqual(variantResult.flag_key, "flag1")
            XCTAssertEqual(variantResult.reason, "MATCH_EVALUATION_REASON")
            XCTAssertTrue(variantResult.segment_keys.contains("segment1"))
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testBoolean() {
        do {
            let booleanResult = try evaluationClient?.evaluateBoolean(
                flagKey: "flag_boolean",
                entityID: "someentity",
                evalContext: ["fizz": "buzz"])
            guard let booleanResult else {
                XCTFail("booleanResult is nil")
                return
            }
            XCTAssertEqual(booleanResult.flag_key, "flag_boolean")
            XCTAssertTrue(booleanResult.enabled)
            XCTAssertEqual(booleanResult.reason, "MATCH_EVALUATION_REASON")
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testBatch() {
        do {
            let requests = [
                EvaluationRequest(flag_key: "flag1", entity_id: "someentity", context: ["fizz": "buzz"]),
                EvaluationRequest(flag_key: "flag_boolean", entity_id: "someentity", context: ["fizz": "buzz"]),
                EvaluationRequest(flag_key: "notfound", entity_id: "someentity", context: ["fizz": "buzz"])
            ]

            let batchResult = try evaluationClient?.evaluateBatch(requests: requests)
            guard let batchResult else {
                XCTFail("batchResult is nil")
                return
            }
            XCTAssertEqual(batchResult.responses.count, 3)

            let variantResult = batchResult.responses[0]
            XCTAssertEqual(variantResult.type, "VARIANT_EVALUATION_RESPONSE_TYPE")
            guard let variantResp = variantResult.variant_evaluation_response else {
                XCTFail("unexpected response type")
                return
            }
            XCTAssertTrue(variantResp.match)
            XCTAssertEqual(variantResp.flag_key, "flag1")
            XCTAssertEqual(variantResp.reason, "MATCH_EVALUATION_REASON")
            XCTAssertTrue(variantResp.segment_keys.contains("segment1"))

            let booleanResult = batchResult.responses[1]
            XCTAssertEqual(booleanResult.type, "BOOLEAN_EVALUATION_RESPONSE_TYPE")
            guard let boolResp = booleanResult.boolean_evaluation_response else {
                XCTFail("unexpected response type")
                return
            }
            XCTAssertEqual(boolResp.flag_key, "flag_boolean")
            XCTAssertTrue(boolResp.enabled)
            XCTAssertEqual(boolResp.reason, "MATCH_EVALUATION_REASON")

            let errorResult = batchResult.responses[2]
            XCTAssertEqual(errorResult.type, "ERROR_EVALUATION_RESPONSE_TYPE")
            guard let errResp = errorResult.error_evaluation_response else {
                XCTFail("unexpected response type")
                return
            }
            XCTAssertEqual(errResp.flag_key, "notfound")
            XCTAssertEqual(errResp.namespace_key, "default")
            XCTAssertEqual(errResp.reason, "NOT_FOUND_ERROR_EVALUATION_REASON")
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testListFlags() {
        do {
            let flags = try evaluationClient?.listFlags()
            guard let flags else {
                XCTFail("flags is nil")
                return
            }
            XCTAssertFalse(flags.isEmpty)
            XCTAssertEqual(flags.count, 2)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testVariantFailure() {
        do {
            _ = try evaluationClient?.evaluateVariant(
                flagKey: "nonexistent",
                entityID: "someentity",
                evalContext: ["fizz": "buzz"])
            XCTFail("Expected an error, but got none")
        } catch let error as FliptClient.ClientError {
            XCTAssertFalse(error.localizedDescription.isEmpty)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testSetGetSnapshotWithInvalidFliptURL() {
        do {
            let snapshot = try evaluationClient?.getSnapshot()
            XCTAssertNotNil(snapshot)

            let invalidFliptClient = try FliptClient(
                namespace: "default",
                url: "http://invalid.flipt.com",
                errorStrategy: .fallback,
                snapshot: snapshot)

            let context = ["fizz": "buzz"]

            for _ in 0 ..< 5 {
                // Variant evaluation
                let variant = try invalidFliptClient.evaluateVariant(
                    flagKey: "flag1",
                    entityID: "entity",
                    evalContext: context)
                XCTAssertEqual(variant.flag_key, "flag1")
                XCTAssertTrue(variant.match)
                XCTAssertEqual(variant.reason, "MATCH_EVALUATION_REASON")
                XCTAssertEqual(variant.variant_key, "variant1")
                XCTAssertTrue(variant.segment_keys.contains("segment1"))

                // Boolean evaluation
                let boolean = try invalidFliptClient.evaluateBoolean(
                    flagKey: "flag_boolean",
                    entityID: "entity",
                    evalContext: context)
                XCTAssertEqual(boolean.flag_key, "flag_boolean")
                XCTAssertTrue(boolean.enabled)
                XCTAssertEqual(boolean.reason, "MATCH_EVALUATION_REASON")

                // List flags
                let flags = try invalidFliptClient.listFlags()
                XCTAssertEqual(flags.count, 2)

                // Get snapshot
                let snapshot = try invalidFliptClient.getSnapshot()
                XCTAssertNotNil(snapshot)
            }

            invalidFliptClient.close()
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testTlsConfigSerialization() {
        do {
            let tlsConfig = try TlsConfig.builder()
                .caCertData("-----BEGIN CERTIFICATE-----")
                .insecureSkipVerify(true)
                .insecureSkipHostnameVerify(true)
                .clientCertData("-----BEGIN CERTIFICATE-----")
                .clientKeyData("-----BEGIN PRIVATE KEY-----")
                .build()

        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase

        do {
            let jsonData = try encoder.encode(tlsConfig)
            guard let jsonString = String(data: jsonData, encoding: .utf8) else {
                XCTFail("Failed to convert data to string")
                return
            }

            // Verify snake_case field names in JSON
            XCTAssertTrue(jsonString.contains("\"ca_cert_data\""))
            XCTAssertTrue(jsonString.contains("\"insecure_skip_verify\""))
            XCTAssertTrue(jsonString.contains("\"insecure_skip_hostname_verify\""))
            XCTAssertTrue(jsonString.contains("\"client_cert_data\""))
            XCTAssertTrue(jsonString.contains("\"client_key_data\""))

            // Verify values are correctly encoded
            XCTAssertTrue(jsonString.contains("\"-----BEGIN CERTIFICATE-----\""))
            XCTAssertTrue(jsonString.contains("true"))
            XCTAssertTrue(jsonString.contains("\"-----BEGIN PRIVATE KEY-----\""))

            // Verify file fields are not present (since we only used data fields)
            XCTAssertFalse(jsonString.contains("ca_cert_file"))
            XCTAssertFalse(jsonString.contains("client_cert_file"))
            XCTAssertFalse(jsonString.contains("client_key_file"))
        } catch {
            XCTFail("Failed to encode TlsConfig: \(error)")
        }
        } catch {
            XCTFail("Failed to create TlsConfig: \(error)")
        }
    }

    func testTlsConfigSerializationWithNilFields() {
        do {
            let tlsConfig = try TlsConfig.builder()
                .insecureSkipHostnameVerify(true)
                .build()

        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase

        do {
            let jsonData = try encoder.encode(tlsConfig)
            guard let jsonString = String(data: jsonData, encoding: .utf8) else {
                XCTFail("Failed to convert data to string")
                return
            }

            // Should only contain the field that was set
            XCTAssertTrue(jsonString.contains("\"insecure_skip_hostname_verify\":true"))

            // Should not contain null/nil fields
            XCTAssertFalse(jsonString.contains("ca_cert_file"))
            XCTAssertFalse(jsonString.contains("ca_cert_data"))
            XCTAssertFalse(jsonString.contains("insecure_skip_verify"))
            XCTAssertFalse(jsonString.contains("client_cert_file"))
            XCTAssertFalse(jsonString.contains("client_key_file"))
            XCTAssertFalse(jsonString.contains("client_cert_data"))
            XCTAssertFalse(jsonString.contains("client_key_data"))
        } catch {
            XCTFail("Failed to encode TlsConfig: \(error)")
        }
        } catch {
            XCTFail("Failed to create TlsConfig: \(error)")
        }
    }
}

File: flipt-client-csharp/tests/FliptClient.Tests/FliptClientTests.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using Xunit;
using Xunit.Abstractions;
using FliptClient;
using FliptClient.Models;

namespace FliptClient.Tests
{
    public class FliptClientTests : IDisposable
    {
        private readonly FliptClient _client;

        public FliptClientTests(ITestOutputHelper output)
        {
            Console.SetOut(new ConsoleWriter(output));
            string? fliptUrl = Environment.GetEnvironmentVariable("FLIPT_URL");
            string? authToken = Environment.GetEnvironmentVariable("FLIPT_AUTH_TOKEN");

            if (string.IsNullOrEmpty(fliptUrl))
            {
                throw new ValidationException("FLIPT_URL environment variable is not set");
            }

            if (string.IsNullOrEmpty(authToken))
            {
                throw new ValidationException("FLIPT_AUTH_TOKEN environment variable is not set");
            }

            var options = new ClientOptions
            {
                Url = fliptUrl,
                Namespace = "default",
                Authentication = new Authentication { ClientToken = authToken }
            };

            // Configure TLS if HTTPS URL is provided
            if (fliptUrl.StartsWith("https://"))
            {
                string? caCertPath = Environment.GetEnvironmentVariable("FLIPT_CA_CERT_PATH");
                if (!string.IsNullOrEmpty(caCertPath))
                {
                    options.TlsConfig = TlsConfig.WithCaCertFile(caCertPath);
                }
                else
                {
                    // Fallback to insecure for local testing
                    options.TlsConfig = TlsConfig.Insecure();
                }
            }

            _client = new FliptClient(options);
        }

        public class ConsoleWriter : StringWriter
        {
            private readonly ITestOutputHelper _output;
            public ConsoleWriter(ITestOutputHelper output)
            {
                _output = output;
            }

            public override void WriteLine(string? message)
            {
                if (message != null)
                {
                    _output.WriteLine(message);
                }
            }
        }

        public void Dispose()
        {
            _client.Dispose();
        }

        [Fact]
        public void TestNullFlagKey()
        {
            Assert.Throws<ValidationException>(() => _client.EvaluateBoolean(null!, "someentity", new Dictionary<string, string>()));
        }

        [Fact]
        public void TestEmptyFlagKey()
        {
            Assert.Throws<ValidationException>(() => _client.EvaluateBoolean("", "someentity", new Dictionary<string, string>()));
        }

        [Fact]
        public void TestNullEntityId()
        {
            Assert.Throws<ValidationException>(() => _client.EvaluateBoolean("flag1", null!, new Dictionary<string, string>()));
        }

        [Fact]
        public void TestEmptyEntityId()
        {
            Assert.Throws<ValidationException>(() => _client.EvaluateBoolean("flag1", "", new Dictionary<string, string>()));
        }

        [Fact]
        public void TestEvaluateVariant()
        {
            var context = new Dictionary<string, string> { { "fizz", "buzz" } };
            var response = _client.EvaluateVariant("flag1", "someentity", context)!;

            Assert.Equal("flag1", response.FlagKey);
            Assert.True(response.Match);
            Assert.Equal("MATCH_EVALUATION_REASON", response.Reason);
            Assert.Equal("variant1", response.VariantKey);
            Assert.Contains("segment1", response.SegmentKeys);
        }

        [Fact]
        public void TestEvaluateBoolean()
        {
            var context = new Dictionary<string, string> { { "fizz", "buzz" } };
            var response = _client.EvaluateBoolean("flag_boolean", "someentity", context)!;

            Assert.Equal("flag_boolean", response.FlagKey);
            Assert.True(response.Enabled);
            Assert.Equal("MATCH_EVALUATION_REASON", response.Reason);
        }

        [Fact]
        public void TestEvaluateBatch()
        {
            var requests = new List<EvaluationRequest>
            {
                new EvaluationRequest { FlagKey = "flag1", EntityId = "someentity", Context = new Dictionary<string, string> { { "fizz", "buzz" } } },
                new EvaluationRequest { FlagKey = "flag_boolean", EntityId = "someentity", Context = new Dictionary<string, string> { { "fizz", "buzz" } } },
                new EvaluationRequest { FlagKey = "notfound", EntityId = "someentity", Context = new Dictionary<string, string> { { "fizz", "buzz" } } }
            };

            var result = _client.EvaluateBatch(requests)!;
            Assert.NotNull(result);
            Assert.Equal(3, result.Responses.Count);

            var variantResponse = result.Responses[0];
            Assert.Equal("VARIANT_EVALUATION_RESPONSE_TYPE", variantResponse.Type);
            Assert.NotNull(variantResponse.VariantEvaluationResponse);
            Assert.Equal("flag1", variantResponse.VariantEvaluationResponse.FlagKey);
            Assert.True(variantResponse.VariantEvaluationResponse.Match);
            Assert.Equal("MATCH_EVALUATION_REASON", variantResponse.VariantEvaluationResponse.Reason);
            Assert.Equal("variant1", variantResponse.VariantEvaluationResponse.VariantKey);
            Assert.Contains("segment1", variantResponse.VariantEvaluationResponse.SegmentKeys);

            var booleanResponse = result.Responses[1];
            Assert.Equal("BOOLEAN_EVALUATION_RESPONSE_TYPE", booleanResponse.Type);
            Assert.NotNull(booleanResponse.BooleanEvaluationResponse);
            Assert.Equal("flag_boolean", booleanResponse.BooleanEvaluationResponse.FlagKey);
            Assert.True(booleanResponse.BooleanEvaluationResponse.Enabled);
            Assert.Equal("MATCH_EVALUATION_REASON", booleanResponse.BooleanEvaluationResponse.Reason);

            var errorResponse = result.Responses[2];
            Assert.Equal("ERROR_EVALUATION_RESPONSE_TYPE", errorResponse.Type);
            Assert.NotNull(errorResponse.ErrorEvaluationResponse);
            Assert.Equal("notfound", errorResponse.ErrorEvaluationResponse.FlagKey);
            Assert.Equal("default", errorResponse.ErrorEvaluationResponse.NamespaceKey);
            Assert.Equal("NOT_FOUND_ERROR_EVALUATION_REASON", errorResponse.ErrorEvaluationResponse.Reason);
        }

        [Fact]
        public void TestEvaluateVariantFailure()
        {
            var context = new Dictionary<string, string> { { "fizz", "buzz" } };
            var exception = Assert.Throws<EvaluationException>(() => _client.EvaluateVariant("nonexistent", "someentity", context));
            Assert.Equal("invalid request: failed to get flag information default/nonexistent", exception.Message);
        }

        [Fact]
        public void TestEvaluateBooleanFailure()
        {
            var context = new Dictionary<string, string> { { "fizz", "buzz" } };
            var exception = Assert.Throws<EvaluationException>(() => _client.EvaluateBoolean("nonexistent", "someentity", context));
            Assert.Equal("invalid request: failed to get flag information default/nonexistent", exception.Message);
        }

        [Fact]
        public void TestListFlags()
        {
            var flags = _client.ListFlags()!;
            Assert.NotEmpty(flags);
            Assert.Equal(2, flags.Length);
        }

        [Fact]
        public void TestClientOptions()
        {
            var options = new ClientOptions { ErrorStrategy = ErrorStrategy.Fallback };
            string optsJson = JsonSerializer.Serialize(options);
            string expectedJson = """{"environment":"default","namespace":"default","url":"http://localhost:8080","error_strategy":"fallback"}""";
            Assert.Equal(expectedJson, optsJson);
        }

        [Fact]
        public void TestSetGetSnapshotWithInvalidFliptURL()
        {
            // Get a snapshot from a working client
            string? snapshot = _client.GetSnapshot();
            Assert.False(string.IsNullOrEmpty(snapshot));

            // Create a client with the previous snapshot and an invalid URL
            var authToken = Environment.GetEnvironmentVariable("FLIPT_AUTH_TOKEN") ?? throw new ValidationException("FLIPT_AUTH_TOKEN environment variable is not set");
            var options = new ClientOptions
            {
                Url = "http://invalid.flipt.com",
                Namespace = "default",
                Snapshot = snapshot!,
                ErrorStrategy = ErrorStrategy.Fallback,
                Authentication = new Authentication { ClientToken = authToken }
            };
            using var invalidClient = new FliptClient(options);

            var context = new Dictionary<string, string> { { "fizz", "buzz" } };

            for (int i = 0; i < 5; i++)
            {
                var variantResponse = invalidClient.EvaluateVariant("flag1", "someentity", context);
                Assert.NotNull(variantResponse);
                Assert.Equal("flag1", variantResponse!.FlagKey);
                Assert.True(variantResponse.Match);
                Assert.Equal("MATCH_EVALUATION_REASON", variantResponse.Reason);
                Assert.Equal("variant1", variantResponse.VariantKey);
                Assert.Contains("segment1", variantResponse.SegmentKeys);

                var booleanResponse = invalidClient.EvaluateBoolean("flag_boolean", "someentity", context);
                Assert.NotNull(booleanResponse);
                Assert.Equal("flag_boolean", booleanResponse!.FlagKey);
                Assert.True(booleanResponse.Enabled);
                Assert.Equal("MATCH_EVALUATION_REASON", booleanResponse.Reason);

                var flags = invalidClient.ListFlags();
                Assert.NotNull(flags);
                Assert.Equal(2, flags!.Length);

                snapshot = invalidClient.GetSnapshot();
                Assert.False(string.IsNullOrEmpty(snapshot));
            }
        }
    }
}
